File: app/src/main/java/cz/feldis/gasprices/utils/RegressionCalculator.kt
package cz.feldis.gasprices.utils

import com.github.mikephil.charting.data.Entry

object RegressionCalculator {

    data class LinearRegressionResult(val slope: Float, val intercept: Float)

    fun calculateSlopeAndIntercept(dataPoints: List<Entry>): LinearRegressionResult {
        val n = dataPoints.size
        if (n == 0) return LinearRegressionResult(0f, 0f)

        val sumX = dataPoints.fold(0f) { sum, entry -> sum + entry.x }
        val sumY = dataPoints.fold(0f) { sum, entry -> sum + entry.y }
        val sumXX = dataPoints.fold(0f) { sum, entry -> sum + entry.x * entry.x }
        val sumXY = dataPoints.fold(0f) { sum, entry -> sum + entry.x * entry.y }

        val denominator = n * sumXX - sumX * sumX
        
        val slope = if (denominator != 0f) (n * sumXY - sumX * sumY) / denominator else 0f
        val intercept = (sumY - slope * sumX) / n

        return LinearRegressionResult(slope, intercept)
    }
}

File: ios/GasPricesiOS/Domain/RegressionCalculator.swift
import Foundation

struct PriceEntry: Equatable {
    let x: Double
    let y: Double
}

struct LinearRegressionResult: Equatable {
    let slope: Double
    let intercept: Double
}

enum RegressionCalculator {
    static func calculateSlopeAndIntercept(dataPoints: [PriceEntry]) -> LinearRegressionResult {
        let n = Double(dataPoints.count)
        guard n > 0 else {
            return LinearRegressionResult(slope: 0, intercept: 0)
        }

        let sumX = dataPoints.reduce(0) { $0 + $1.x }
        let sumY = dataPoints.reduce(0) { $0 + $1.y }
        let sumXX = dataPoints.reduce(0) { $0 + ($1.x * $1.x) }
        let sumXY = dataPoints.reduce(0) { $0 + ($1.x * $1.y) }

        let denominator = n * sumXX - sumX * sumX
        let slope = denominator != 0 ? (n * sumXY - sumX * sumY) / denominator : 0
        let intercept = (sumY - slope * sumX) / n

        return LinearRegressionResult(slope: slope, intercept: intercept)
    }
}